scli/scli.1
.\"                              hey, Emacs:   -*- nroff -*-
.\" stop is free software; you can redistribute it and/or modify
.\" it under the terms of the GNU General Public License as published by
.\" the Free Software Foundation; either version 2 of the License, or
.\" (at your option) any later version.
.\"
.\" This program is distributed in the hope that it will be useful,
.\" but WITHOUT ANY WARRANTY; without even the implied warranty of
.\" MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
.\" GNU General Public License for more details.
.\"
.\" You should have received a copy of the GNU General Public License
.\" along with this program; see the file COPYING.  If not, write to
.\" the Free Software Foundation, 675 Mass Ave, Cambridge, MA 02139, USA.
.\"
.TH SCLI 1 "August 26, 2001"
.\" Please update the above date whenever this man page is modified.
.\"
.\" Some roff macros, for reference:
.\" .nh        disable hyphenation
.\" .hy        enable hyphenation
.\" .ad l      left justify
.\" .ad b      justify to both left and right margins (default)
.\" .nf        disable filling
.\" .fi        enable filling
.\" .br        insert line break
.\" .sp <n>    insert n+1 empty lines
.\" for manpage-specific macros, see man(7)
.SH NAME
scli \- SNMP Command Line Interface
.SH SYNOPSIS
.B scli
.RI [ options ]
.I [hostname]
.RI [ community ]
.SH DESCRIPTION
\fBscli\fP provides a simple command line interface on top of the
Simple Network Management Protocol (SNMP). It can be used to read
data from devices and to configure them with simple commands. It
relies on the capabilities of the devices and thus not all commands
may work with all devices.
.PP
The optional \fIhostname\fR argument is the name or the IP address of
the remote host. \fBscli\fP talks to this host's SNMP agent on the
default port number 161 using the SNMP over UDP transport mapping.
.PP
The optional \fIcommunity\fR argument is the community string used to
communicate with the remote SNMP agent. The default string is
\fIpublic\fR. Please check the configuration of the SNMP agent to get
a clue about the community string to use, if the default community
string does not work.
.PP
.SH OPTIONS
\fBscli\fP accepts the following options:
.TP
.B \-V, \-\-version
Show version of program.
.TP
.B \-c, \-\-command
Process the given commands and exit.
.TP
.B \-d seconds, \-\-delay=seconds
Set delay between screen updates in seconds. The default delay between
updates is 5 seconds.
.TP
.B \-f, \-\-file
Process commands from a file and exit.
.TP
.B \-h, \-\-help
Show summary of scli options.
.TP
.B \-n, \-\-norc
Do not evaluate ~/.sclirc on startup.
.TP
.B \-p number, \-\-port=number
Set the SNMP port number. The default port number is 161.
.TP
.B \-r count, \-\-retries=count
The number of SNMP retransmissions before we give up. The default
number of retries is 5.
.TP
.B \-t msec, \-\-timeout=msec
The initial timeout in milli-seconds before an SNMP request is
transmitted. The default initial timeout interval is 200 milli-seconds.
.TP
.B \-x, \-\-xml
Generate XML output instead of the default output which is optimized
for humans and more difficult to parse by programs.
.SH COMMAND OVERVIEW
\fBscli\fP is a simple command interpreter. \fBscli\fP commands are
organized in a hierarchy. This section only describes the top-level
commands. Invoke the \fIhelp\fP command described below to get a
complete list of all commands supported by \fBscli\fP.
.TP 10
.B open
Establish an association to a remote SNMP agent.
.TP
.B close
Close the association to a remote SNMP agent.
.TP
.B exit
Exit the \fBscli\fP command interpreter.
.TP
.B help
Display help information about the \fBscli\fP command interpreter
including a full list of the \fBscli\fP command hierarchy.
.TP
.B history
Show the history of the last \fBscli\fP commands.
.TP
.B create
Create new object instances on the remote SNMP agent. This command
has many sub-commands which are used to instantiate very different
things.
.TP
.B delete
Delete object instances from the remote SNMP agent. This command
has many sub-commands which are used to delete very different
things.
.TP
.B set
Modify object instances on the remote SNMP agent by assigning new
values. This command has many sub-commands which are used to set
very different things.
.TP
.B show
Show information provided by the remote SNMP agent. This command has
many sub-commands focusing on various aspects of the system.
.TP
.B monitor
Monitor information provided by the remote SNMP agent. This command
has many sub-commands focusing on various aspects of the system.
\fBscli\fP monitors provide a live display of network and device
activities. Information is shown in a compact human readable format
and updated in regular intervals. The initial update interval length
is determined by the delay command line option.
.PP
\fBscli\fP supports recursive command evaluation. When a command is
entered which is not a leaf of the command tree, then the interpreter
will recursively invoke all commands in the subtree identified by the
entered command. In particular, entering \fIshow\fR will cause
\fBscli\fP to retrieve and display all information accessible from the
remote SNMP agent.

.SH SYSTEM MODE
The system mode provides information about general system capabilities
and it provides core system configuration commands. It is primarily
based on the SNMPv2-MIB as published in RFC 1907, the
HOST-RESOURCES-MIB as published in RFC 2790, and the ENTITY-MIB as
published in RFC 2737. Some commands which probe for certain features
rely on additional MIBs.
.TP
.B set system name <string>
.TP
.B set system contact <string>
.TP
.B set system location <string>
.TP
.B show system info
.TP
.B show system devices
.TP
.B show system storage
.TP
.B show system mounts
.TP
.B show system processes
.TP
.B show system entity details
.TP
.B show system entity containment
.TP
.B monitor system storage
.TP
.B monitor system processes

.SH INTERFACE MODE
The interface mode is based on the IF-MIB as published in RFC 2863 and
provides the following commands:
.TP
.B show interface info [<regexp>]
.TP
.B show interface details [<regexp>]
.TP
.B show interface stats [<regexp>]
.TP
.B show interface stack
.TP
.B set interface status <regexp> <value>
.TP
.B monitor interface stats [<regexp>]
.P
The \fBshow interface info\fP command shows summary information about
all selected interfaces. The optional regular expression <regexp> is
matched against the interface descriptions to select the interfaces of
interest.  The command generates a table which displays the interface
number, its status (administrative status, operational status,
connector present, promiscuous mode), the maximum transfer unit (MTU)
of the interface, the interface type and speed, the interface's local
name and its description.
.P
The \fBshow interface details\fP command describes the selected
interfaces in more detail. The optional regular expression <regexp> is
matched against the interface descriptions to select the interfaces of
interest. <tbd>
.P
The \fBshow interface stats\fP command <tbd>
.P
The \fBshow interface stack\fP command shows the stacking order of the
interfaces. <tbd>
.P
THe \fBset interface status <regexp> <value>\fP command modifies the
administrative status of all selected interfaces. The optional regular
expression <regexp> is matched against the interface descriptions to
select the interfaces of interest. The <value> parameter must be one
of the strings "up", "down" or "testing".

.SH NORTEL MODE
The nortel mode allows to manipulate virtual lans (vlans) on nortel
bridges. It is based on the RAPIDCITY-VLAN-MIB which is implemented at
least on the baystack bridges.
.TP
.B create nortel bridge vlan <vlanid> <name>
.TP
.B delete nortel bridge vlan <regexp>
.TP
.B set nortel bridge vlan ports <regexp> <ports>
.TP
.B show nortel bridge vlan info [<regexp>]
.TP
.B show nortel bridge vlan details [<regexp>]
.P
The \fBcreate nortel bridge vlan\fP command is used to create a new
vlan with the given vlanid and name.
.P
The \fBdelete nortel bridge vlan\fP command deletes all selected
vlans.  The regular expression <regexp> is matched against the vlan
names to select the vlans that should be deleted.
.P
The \fBset nortel bridge vlan ports\fP command allows to assign ports
to port-based vlans. The regular expression <regexp> is matched
against the vlan names to select the vlans that should be
modified. The <ports> argument contains a comma separated list of port
numbers or port number ranges, e.g. 1,5,7-8.
.P
The \fBshow nortel bridge vlan info\fP command shows summary
information about all selected vlans.  The optional regular expression
<regexp> is matched against the vlan names to select the vlans of
interest. The command generates a table which displays the vlan
number, the vlan name, the vlan type, the assigned vlan color, the
vlan priority, whether routing is enabled and the status of the vlan.
.P
The \fBshow nortel bridge vlan details\fP command describes the
selected vlans in more detail. The optional regular expression
<regexp> is matched against the vlan names to select the vlans of
interest. <tbd>

.SH SCLI MODE
The scli mode can be used to inspect and modify the status of the scli
interpreter.
.TP
.B set scli debugging [<layer> [<layer> [...]]]
.TP
.B set scli pager <string>
.TP
.B set scli format <fmt>
.TP
.B create scli alias <name> <value>
.TP
.B delete scli alias <name>
.TP
.B show scli aliases
.TP
.B show scli info
.TP
.B show scli modes

.SH INTERACTIVE MONITOR COMMANDS
Several single-key commands are recognized while \fBscli\fP is running
in monitor mode:
.TP 8
.B space
Update the display.
.TP
.B ^L
Redraw the screen.
.TP
.B h or ?
Print a help page.
.TP
.B c
Toggle display of contact summary information.
.TP
.B d
Set the delay in seconds between updates.
.TP
.B m
Toggle display of mode specific summary information.
.TP
.B n
Toggle display of network layer summary information.
.TP
.B q
Quit the monitoring mode.
.TP
.B x
Exit scli immediately.
.TP
.B t
Toggle display of transport layer summary information.
.TP
.B w
Freeze the screen until someone hits a key.
.TP
.B >
Select next mode in the list of display modes.
.TP
.B <
Select previous mode in the list of display modes.
.SH UNITS
\fBscli\fP displays numbers in a compact format. It uses several
abbreviations to explain the units of the numbers. Here are some
explanations:
.TP
.B bps
Bits per second.
.TP
.B pps
Packets per second.
.TP
.B sps
Segments per second.
.PP
Large numbers are usually displayed with a scaling factor (k=10^3,
m=10^6, g=10^9, t=10^12, p=10^15, K=2^10, M=2^20, G=2^30, T=2^40m,
P=2^50).
.SH ALIASES
Long command names can be abbreviated by using aliases. The \fBscli\fP
interpreter maintains a list of aliases. Every command is first checked
whether the first word matches an alias. If the first word matches an
alias, then the first word will be replaced by the value of the alias.
Alias expansion only happens once.
.P
Aliases are not expanded when the \fBscli\fP interpreter is running
in non-interactive mode.
.SH REGULAR EXPRESSIONS
\fBscli\fP generally uses POSIX.2 Extended Regular Expressions. The
regular expression matching is case sensitive.
.SH READLINE
\fBscli\fP supports the GNU readline library. The GNU readline library
provides command line editing as well as command completion
capabilities. Please consult the GNU readline documentation for a
detailed description of the readline features.
.SH HISTORY
\fBscli\fP supports the GNU history library. The GNU history library
provides a history expansion feature that is similar to the history
expansion provided by `csh'. Please consult the GNU history library
documentation for a detailed description of the history features.
.SH PAGER
\fBscli\fP respects the PAGER environment variable when running
interactively.  If the PAGER environment variable exists and the
output generated by an scli command does not fit on the terminal, then
the output is written to the standard input of the PAGER.
.SH FILES
.TP
.I ~/.scli_history
The history of scli commands.
.TP
.I ~/.sclirc
The scli startup file.
.TP
.I ~/.inputrc
Individual readline initialization file.
.SH "FURTHER INFORMATION"
More information can be found on the stools web page which is
available at:
.PP
<http://www.ibr.cs.tu-bs.de/projects/stools/>
.SH "SEE ALSO"
.BR regex(7)
.SH AUTHOR
Juergen Schoenwaelder <schoenw@ibr.cs.tu-bs.de>.


NEWS
Changes in stools 0.1.15:

- stop supports a new tcp mode
- scli supports a new "show udp listener" command
- scli supports a new "show tcp listener" command
- scli shows long output in your favorite pager if $PAGER is defined
- stubs now check the SNMP ASN.1 type tag in responses
- scli supports new "open" and "close" commands
- scli shows the remote host name in the prompt
- minor bug fixes

Changes in stools 0.1.14:

- scli supports a new "show ip mapping" command
- scli supports a new bridge mode (info, ports, forwarding, filter)
- stop supports a new disman mode
- scli supports a new "show interface info" command
- scli supports a new "show disman script info" command
- renamed "show interface tunnel" to "show ip tunnel"
- minor bug fixes

Changes in stools 0.1.13:

- scli supports a new "show interface tunnel" command
- scli supports a new "show ip addresses" command
- scli "show interface info" command displays IPv4 address prefixes
- minor bug fixes

Changes in stools 0.1.12:

- improved instance identifier decoding in the stubs
- scli supports a new "show tcp connections" command
- scli supports a new "show disman languages" command
- scli supports a new "show disman scripts" command
- scli "show interface info" command displays IPv4 addresses
- minor bug fixes

Changes in stools 0.1.11:

- recursive command evaluation on non-leaf nodes of the command tree
- stop now displays packet statistics in addition to byte statistics
- scli supports a new "show physical components" command
- scli supports a new "show physical containment" command
- renamed scli command "show system" to "show system info"
- renamed scli command "show interface" to "show interface info"
- minor bug fixes

Changes in stools 0.1.10:

- optimized the size of the generated stub code plus simple index decoding
- scli supports a new "show system processes" command
- work on a disman mode for scli started
- minor bug fixes

Changes in stools 0.1.9:

- added the -c command line option to scli
- added history expansions to scli based on the GNU history library
- rewrote the scli modes so that they do not directly write to stdout
- scli supports a new "show system mounts" command

Changes in stools 0.1.8:

- scli supports a new "show system storage" command
- bug fixes and better portability (fixed some ANSI C violations)
- some improvements in the documentation

Overview of stools 0.1.7:

- first public version of the stools package


scli/scli.c
/* 
 * scli.c -- emulate a command line interface via SNMP
 *
 *       After more than 10 years of SNMP, I felt it is time for the
 *       first really useful command line SNMP management tool. ;-)
 *
 * Copyright (C) 2001 Juergen Schoenwaelder
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2, or (at your option)
 * any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 59 Temple Place - Suite 330, Boston, MA 02111-1307, USA.
 * 
 * @(#) $Id$
 */

#include "scli.h"

#include <getopt.h>
#include <string.h>
#include <errno.h>
#include <sys/types.h>
#include <sys/stat.h>
#include <fcntl.h>
#include <signal.h>
#ifdef HAVE_SYS_IOCTL_H
#include <sys/ioctl.h>
#endif

#include <readline/readline.h>
#include <readline/history.h>

static scli_interp_t *current_interp = NULL;

/*
 * Make sure we quit curses when we receive signals to abort the
 * process. Otherwise, we might leave the terminal in an unusable
 * state.
 */

static void
onsignal(int n)
{
    scli_curses_off();
    exit(n);
}

/*
 * Track window change events using whatever mechanism is available.
 */

static void
onwinch(int n)
{
    scli_get_screen(NULL, NULL);
    (void) signal(SIGWINCH, onwinch);	/* some systems need this */
}

/*
 * This is the main loop of the scli command interpreter. It reads a
 * command and calls the function which has been register to process
 * the command.
 */

static void
mainloop(scli_interp_t *interp)
{
    char *input, *expansion, *prompt;
    int result, len, code = SCLI_OK;
    GString *s;

    g_return_if_fail(interp);

    s = g_string_new(NULL);
    while (code != SCLI_EXIT) {
	current_interp = interp;
	prompt = scli_prompt(interp);
	input = readline(prompt);
	g_free(prompt);
	current_interp = NULL;
 	if (! input) {
	    g_print("\n");
	    code = SCLI_EXIT;
	    break;
	}
	if (! *input) {
	    free(input);
	    continue;
	}
	result = history_expand(input, &expansion);
	if (result < 0 || result == 2) {
	    free(input);
	    free(expansion);
	    continue;
	}
	free(input); input = expansion;
	len = strlen(input);
	if (len > 0
	    && input[len-1] == '\\'
	    && (len == 1 || input[len-2] != '\\')) {
	    input[len-1] = 0;
	    g_string_append(s, input);
	} else {
	    g_string_append(s, input);
	    add_history(s->str);
	    code = scli_eval_string(interp, s->str);
	    g_string_truncate(s, 0);
	}
	free(input);
    }
    g_string_free(s, 1);
}

/*
 * Generator function for command completion.  STATE lets us know
 * whether to start from scratch; without any state (i.e. STATE == 0),
 * then we start at the top of the list.
 */

static char*
generator(char const *text, int state)
{
    static int argc = 0;
    static char **argv = NULL;
    static size_t len;
    static GNode *last_node = NULL;
    static GSList *last_elem = NULL;

    scli_interp_t *interp = current_interp;	/* a hack, for sure */
    GNode *node;
    int i;

    /*
     * Check whether we have a valid command interpreter...
     */

    g_return_val_if_fail(interp, NULL);
    g_return_val_if_fail(interp->cmd_root, NULL);
    
    /*
     * If this is a new word to complete, initialize now.  This
     * includes saving the length of TEXT for efficiency, and
     * initializing the index variable to 0.
     */

    if (! state) {
	if (argv) {
	    g_free(argv);
	    argv = NULL;
	}
	if (scli_split(rl_line_buffer, &argc, &argv) != SCLI_OK) {
	    return NULL;
	}
	len = strlen(text);
	last_node = g_node_first_child(interp->cmd_root);
	for (i = 0; i < (len ? argc-1 : argc); i++) {
	    while (last_node) {
		scli_cmd_t *cmd = (scli_cmd_t *) last_node->data;
		if (strcmp(cmd->name, argv[i]) == 0) {
		    break;
		}
		last_node = g_node_next_sibling(last_node);
	    }
	    if (! last_node) {
		return NULL;
	    }
	    last_node = g_node_first_child(last_node);
	}
	if (argc == 0 || (argc < 2 && len && !isspace((int) text[len-1]))) {
	    last_elem = interp->alias_list;
	} else {
	    last_elem = NULL;
	}
    }

    /*
     * Return the next name which partially matches from the command list.
     */

    for (node = last_node; node; node = g_node_next_sibling(node)) {
	scli_cmd_t *cmd = (scli_cmd_t *) node->data;
	if (strncmp(cmd->name, text, len) == 0) {
	    last_node = g_node_next_sibling(node);
	    return (g_strdup(cmd->name));
	}
    }

    if (! node && last_elem) {
	GSList *elem;
	for (elem = last_elem; elem; elem = g_slist_next(elem)) {
	    scli_alias_t *alias = (scli_alias_t *) elem->data;
	    if (strncmp(alias->name, text, len) == 0) {
		last_elem = g_slist_next(elem);
		return (g_strdup(alias->name));
	    }
	}
    }

    return NULL;
}

/*
 *  Attempt to complete on the contents of TEXT.  START and END bound
 *  the region of rl_line_buffer that contains the word to complete.
 *  TEXT is the word to complete.  We can use the entire contents of
 *  rl_line_buffer in case we want to do some simple parsing.  Return
 *  the array of matches, or NULL if there aren't any.
 */

static char**
completion(char *text, int start, int end)
{
    char **matches;

    matches = (char **)NULL;

    /*
     * If this word is at the start of the line, then it is a command
     * to complete.  Otherwise it is the name of a file in the current
     * directory.
     */

#ifdef HAVE_RL_COMPLETION_MATCHES
    matches = rl_completion_matches(text, generator);
#else
    matches = completion_matches(text, generator);
#endif

    /*
     * Do not switch to filename completion if no matches were found.
     */
    
    if (! matches) {
#ifdef HAVE_RL_COMPLETION_MATCHES
       rl_completion_entry_function = (rl_compentry_func_t *) completion;
#else
       rl_completion_entry_function = (Function *) completion;
#endif
    }
    
    return matches;
}

/*
 * Tell the GNU Readline library how to complete.  We want to try to
 * complete on command names if this is the first word in the line, or
 * on filenames if not.
 */

static void
readline_init()
{
    /* Allow conditional parsing of the ~/.inputrc file. */
    rl_readline_name = "scli";

    /* Tell the completer that we want a crack first. */
    rl_attempted_completion_function = (CPPFunction *) completion;
}

/*
 * Print some usage information if the user fails to call the
 * program correctly.
 */

static void
usage()
{
    g_print("Usage: scli [OPTION] [hostname [community]]\n");
    g_print(
	"Options:\n"
	"  -V, --version     show version information and exit\n"
	"  -c, --command     process the given command and exit\n"
	"  -d, --delay       delay in seconds between screen updates (default 5)\n"
	"  -f, --file        process commands from a file and exit\n"
	"  -h, --help        display this help and exit\n"
	"  -n, --norc        do not evaluate ~/.sclirc on startup\n"
	"  -p, --port        port number of the SNMP agent (default 161)\n"
	"  -r, --retries     number of retries (default 3)\n"
	"  -t, --timeout     timeout between retries in milliseconds (default 500)\n"
	"  -x, --xml         produce machine readable XML output\n"
	"  -y, --dry         parse commands but do not execute them (dry mode)\n"
	);
}

/*
 * The main of scli. Parse arguments, initialize the SNMP session and
 * fall into the mainloop() until we are done.
 */

int
main(int argc, char **argv)
{
    scli_interp_t *interp;
    char *file = NULL;
    char *cmd = NULL;
    int c;
    
    int norc = 0, port = 161, delay = 5000, retries = 3, timeout = 500000;
    int xml = 0, dry = 0;

    static struct option const long_options[] =
    {
        { "version", no_argument,       0, 'V' },
	{ "command", required_argument, 0, 'c' },
        { "delay",   required_argument, 0, 'd' },
	{ "dry",     required_argument, 0, 'y' },
        { "file",    required_argument, 0, 'f' },
        { "help",    no_argument,       0, 'h' },
        { "norc",    no_argument,       0, 'n' },
        { "port",    required_argument, 0, 'p' },
	{ "retries", required_argument, 0, 'r' },
	{ "timeout", required_argument, 0, 't' },
	{ "xml",     no_argument,	0, 'x' },
        { NULL, 0, NULL, 0}
    };

    while ((c = getopt_long(argc, argv, "Vc:d:f:hnp:r:t:xy", long_options,
                            (int *) 0)) != EOF) {
        switch (c) {
        case 'V':
	    g_print("scli version %s\n", VERSION);
	    g_print("Copyright %s\n", scli_copyright);
	    g_print("scli comes with ABSOLUTELY NO WARRANTY.\n"
		    "You may redistribute copies of scli under\n"
		    "the terms of the GNU General Public License.\n");
	    exit(0);
	case 'c':
	    cmd = optarg;
	    break;
        case 'd':
            if (atoi(optarg) > 0) {
                delay = atoi(optarg) * 1000;
            }
            break;
	case 'f':
	    file = optarg;
	    break;
        case 'h':
            usage();
            exit(0);
	case 'n':
	    norc = 1;
	    break;
        case 'p':
            if (atoi(optarg) <= 0 || atoi(optarg) > 65535) {
                port = atoi(optarg);
            }
            break;
	case 'r':
	    if (atoi(optarg) > 0) {
		retries = atoi(optarg);
	    }
	    break;
	case 't':
	    if (atoi(optarg) < 50) {
		timeout = 50 * 1000;
	    } else {
		timeout = atoi(optarg) * 1000;
	    }
	    break;
	case 'x':
	    xml = 1;
	    break;
	case 'y':
	    dry = 1;
	    break;
        default:
            usage();
            exit(1);
        }
    }

    if (argc-optind > 2) {
        usage();
        exit(1);
    }

    /*
     * Call the localtime() function to make sure the timezone
     * variable is set correctly. This is necessary at least on
     * Solaris machines.
     */

    {
	time_t t = 0;
	(void) localtime(&t);
    }

    /*
     * Initialize the g_snmp library.
     */

    if (! g_snmp_init(FALSE)) {
	g_error("scli: Initialization of SNMP library failed");
        exit(1);
    }

    /*
     * Setting up curses in such a way that we are sure to restore
     * the tty into something useful when we are done.
     */

    signal(SIGINT, onsignal);
    signal(SIGTERM, onsignal);
    signal(SIGHUP, onsignal);
    signal(SIGQUIT, onsignal);
    signal(SIGWINCH, onwinch);
    signal(SIGPIPE, SIG_IGN);

    /*
     * Initialize the interpreter. Register the toplevel commands.
     */

    interp = scli_interp_create();

    if (file == NULL && cmd == NULL && isatty(0)) {
	interp->flags |= SCLI_INTERP_FLAG_INTERACTIVE;
    }

    interp->delay = delay;

    if (xml) {
	interp->flags |= SCLI_INTERP_FLAG_XML;
    }

    if (dry) {
	interp->flags |= SCLI_INTERP_FLAG_DRY;
    }

    if (scli_interp_interactive(interp)) {
	g_print("scli version %s %s\n", VERSION, scli_copyright);
    }

    if (scli_interp_interactive(interp)) {
	if (scli_set_pager(interp, getenv("PAGER")) < 0) {
	    g_warning("$PAGER contains unsafe characters - ignored\n");
	}
    }

    if (scli_cmd_open(interp, argc-optind+1, argv+optind-1) != SCLI_OK) {
	g_printerr(interp->result->str);
    }

    scli_init_scli_mode(interp);
    scli_init_system_mode(interp);
    scli_init_entity_mode(interp);
    scli_init_disman_mode(interp);
    scli_init_snmp_mode(interp);
    scli_init_ip_mode(interp);
    scli_init_udp_mode(interp);
    scli_init_tcp_mode(interp);
    scli_init_interface_mode(interp);
    scli_init_bridge_mode(interp);
    scli_init_atm_mode(interp);
    scli_init_ether_mode(interp);
    scli_init_ospf_mode(interp);
    scli_init_printer_mode(interp);
    scli_init_cisco_mode(interp);
    scli_init_nortel_mode(interp);

    if (! norc) {
	(void) scli_eval_init_file(interp);
    }

    if (scli_interp_interactive(interp)) {
        gchar *home = g_get_home_dir();
	gchar *path = NULL;
	readline_init();
	using_history();
	if (home) {
	    path = g_strdup_printf("%s/.scli_history", home);
	    if (access(path, R_OK) == 0) {
		if (read_history(path) != 0) {
		    g_printerr("scli: reading history failed: %s",
			       g_strerror(errno));
		}
	    }
	}
	mainloop(interp);
	if (path) {
	    if (access(path, W_OK) != 0) {
		close(creat(path, 0600));
	    }
	    if (access(path, W_OK) == 0) {
		int history_file_size = 1000;	/* arbitrary choice */
		char *histfilesize = getenv("HISTFILESIZE");
		if (histfilesize) {
		    history_file_size = atoi(histfilesize);
		}
		stifle_history(history_file_size);
		if (write_history(path) != 0) {
		    g_printerr("scli: writing history failed: %s",
			       g_strerror(errno));
		}
	    }
	    g_free(path);
	}
    } else if (file) {
	(void) scli_eval_file(interp, file);
    } else if (cmd) {
	(void) scli_eval_string(interp, cmd);
    } else {
	(void) scli_eval_file_stream(interp, stdin);
    }

    return 0;
}
